File: src/com/xw/vvtts/utils/VoiceProfile.java
package com.xw.vvtts.utils;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * 发音角色配置。
 * 预置 8 个 Kona 角色，支持按角色自定义覆盖 8 个 ECI voice 参数。
 * 默认值来自 KonaVoice（KonaVoicePresets.csv 权威数据），
 * 用户自定义值存 SharedPreferences，恢复默认即删除覆盖。
 */
public class VoiceProfile {
    private static final String PREFS = "vvtts_voice_profile";
    private static final String KEY_PRESET = "preset";   // 当前选中的预置角色 1-8

    // ECI voice param 权威编号（eci.h）
    public static final int PARAM_GENDER = 0;        // 性别 0=男 1=女（二选一）
    public static final int PARAM_HEAD_SIZE = 1;     // 头部大小 0-100
    public static final int PARAM_PITCH_BASE = 2;    // 音调基线 40-120
    public static final int PARAM_PITCH_FLUC = 3;    // 抖动 0-100
    public static final int PARAM_ROUGHNESS = 4;     // 粗糙度 0-100
    public static final int PARAM_BREATHINESS = 5;   // 沙哑度/气声 0-100
    public static final int PARAM_SPEED = 6;         // 语速 0-250
    public static final int PARAM_VOLUME = 7;        // 音量 0-100

    // 弹编辑对话框时展示的音色滑杆参数（gender 单独单选做官方原厂切换；
    // pitchBase/speed/volume 由主界面的音调/语速/音量控制，编辑界面不重复暴露）
    public static final int[] EDITABLE_PARAMS = {
        PARAM_HEAD_SIZE, PARAM_PITCH_FLUC, PARAM_ROUGHNESS, PARAM_BREATHINESS,
    };

    // 预置角色英文名（苹果 Kona 8 角色）
    public static final String[] PRESET_NAMES = {
        "Reed", "Shelley", "Sandy", "Rocko", "Flo", "Grandma", "Grandpa", "Eddy",
    };

    public static final String[] PRESET_NAMES_CN = {
        "Reed", "Shelley", "Sandy", "Rocko", "Flo", "Grandma", "Grandpa", "Eddy",
    };

    private final SharedPreferences prefs;

    public VoiceProfile(Context ctx) {
        prefs = ctx.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public int getPreset() { return prefs.getInt(KEY_PRESET, 1); }

    public void setPreset(int n) {
        if (n < 1) n = 1;
        if (n > 8) n = 8;
        prefs.edit().putInt(KEY_PRESET, n).commit();
    }

    /** 某角色某参数是否有自定义覆盖 */
    public boolean hasOverride(int preset, int param) {
        return prefs.contains(overrideKey(preset, param));
    }

    /** 取某角色的参数值：优先自定义覆盖，否则回落 KonaVoice 默认 */
    public int getParam(int preset, int param) {
        String key = overrideKey(preset, param);
        if (prefs.contains(key)) return prefs.getInt(key, 0);
        return KonaVoice.byPreset(preset).param(param);
    }

    /** 设置某角色的参数自定义覆盖 */
    public void setParam(int preset, int param, int value) {
        prefs.edit().putInt(overrideKey(preset, param), value).commit();
    }

    /** 恢复某角色的默认（删除该角色全部自定义覆盖） */
    public void resetPreset(int preset) {
        SharedPreferences.Editor e = prefs.edit();
        for (int p = 0; p < 8; p++) {
            e.remove(overrideKey(preset, p));
        }
        e.commit();
    }

    private static String overrideKey(int preset, int param) {
        return "override_" + preset + "_" + param;
    }

    /** 参数中文名 */
    public static String paramName(int param) {
        switch (param) {
            case PARAM_GENDER: return "性别";
            case PARAM_HEAD_SIZE: return "头部大小";
            case PARAM_PITCH_BASE: return "音调";
            case PARAM_PITCH_FLUC: return "情感起伏";
            case PARAM_ROUGHNESS: return "粗糙度";
            case PARAM_BREATHINESS: return "沙哑度";
            case PARAM_SPEED: return "语速";
            case PARAM_VOLUME: return "音量";
            default: return "参数" + param;
        }
    }
}

File: src/com/xw/vvtts/utils/VoiceProfile.kt
package com.xw.vvtts.utils

import android.content.Context
import android.content.SharedPreferences

/**
 * Voice profile configuration.
 * 8 preset Kona voices;; per-role custom overrides of the 8 ECI voice params.
 * Defaults come from KonaVoice (KonaVoicePresets.csv is the source of truth),
 * user custom values live in SharedPreferences;; reset removes the overrides.
 */
class VoiceProfile(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    val preset: Int
        get() = prefs.getInt(KEY_PRESET, 1)

    fun setPreset(n: Int) {
        var v = n
        if (v < 1) v = 1
        if (v > 8) v = 8
        prefs.edit().putInt(KEY_PRESET, v).commit()
    }

    /** Whether the given param has a custom override for this preset */
    fun hasOverride(preset: Int, param: Int): Boolean {
        return prefs.contains(overrideKey(preset, param))
    }

    /** Get a preset's param value: custom override first, else KonaVoice default */
    fun getParam(preset: Int, param: Int): Int {
        val key = overrideKey(preset, param)
        if (prefs.contains(key)) return prefs.getInt(key, 0)
        return KonaVoice.byPreset(preset).param(param)
    }

    /** Set a custom override for a preset param */
    fun setParam(preset: Int, param: Int, value: Int) {
        prefs.edit().putInt(overrideKey(preset, param), value).commit()
    }

    /** Restore a preset's defaults (delete all its custom overrides) */
    fun resetPreset(preset: Int) {
        val e = prefs.edit()
        for (p in 0 until 8) {
            e.remove(overrideKey(preset, p))
        }
        e.commit()
    }

    companion object {
        private const val PREFS = "vvtts_voice_profile"
        private const val KEY_PRESET = "preset"   // Currently selected preset role 1-8

        // ECI voice param canonical ids (eci.h)
        const val PARAM_GENDER = 0        // Gender:  0=male  1=female (pick one)
        const val PARAM_HEAD_SIZE = 1     // Head size  0-100
        const val PARAM_PITCH_BASE = 2    // Pitch baseline  40-120
        const val PARAM_PITCH_FLUC =  3    // Pitch flutter  0-100
        const val PARAM_ROUGHNESS = 4     // Roughness  0-100
        const val PARAM_BREATHINESS =  5   // Breathiness  0-100
        const val PARAM_SPEED =  6         // Speed  0-250
        const val PARAM_VOLUME =  7        // Volume  0-100

        // Sliders shown in the edit dialog (gender is a separate radio — official voice switch;
        // pitchBase/speed/volume are controlled from the main screen, not duplicated here)
        val EDITABLE_PARAMS = intArrayOf(
            PARAM_HEAD_SIZE, PARAM_PITCH_FLUC, PARAM_ROUGHNESS, PARAM_BREATHINESS,
        )

        // Preset voice names (Apple Kona 8 roles)
        val PRESET_NAMES = arrayOf(
            "Reed", "Shelley", "Sandy", "Rocko", "Flo", "Grandma", "Grandpa", "Eddy",
        )

        private fun overrideKey(preset: Int, param: Int): String {
            return "override_$preset" + "_" + param
        }

        /** Parameter display name */
        fun paramName(param: Int): String {
            return when (param) {
                PARAM_GENDER -> "Gender"
                PARAM_HEAD_SIZE -> "Head size"
                PARAM_PITCH_BASE -> "Pitch"
                PARAM_PITCH_FLUC -> "Pitch flutter"
                PARAM_ROUGHNESS -> "Roughness"
                PARAM_BREATHINESS -> "Breathiness"
                PARAM_SPEED -> "Speed"
                PARAM_VOLUME -> "Volume"
                else -> "Parameter $param"
            }
        }
    }
}